File: Sentinel/iotjava/iotrmi/Java/basics/TestClassCallbacks_Stub.java
import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;
import iotruntime.master.CommunicationHandler;

public class TestClassCallbacks_Stub {

	public static void main(String[] args) throws Exception {

		CommunicationHandler comHan = new CommunicationHandler(true);
		int numOfPorts = 2;
		//int[] ports = comHan.getCallbackPorts(numOfPorts);

		int localportsend = 5011;
		int localportrecv = 6011;
		int portsend = 5000;
		int portrecv = 6000;
		//String address = "localhost";
		//String address = "192.168.2.191";	// RPi2
		//String skeletonAddress = "128.195.136.170";	// dc-9.calit2.uci.edu
		String skeletonAddress = "128.195.204.132";
		String callbackAddress = "128.195.204.132";	// dw-2.eecs.uci.edu (this machine)
		//String skeletonAddress = "192.168.2.108";	// RPi1
		//String callbackAddress = "192.168.2.191";	// RPi2
		int rev = 0;

		TestClassComplete_Stub tcstub = new TestClassComplete_Stub(localportsend, localportrecv, portsend, portrecv, 
			skeletonAddress, rev);
		System.out.println("==== CALLBACKS ====");
		CallBackInterface cbSingle = new CallBack(2354);
		tcstub.registerCallback(cbSingle);
		System.out.println("Registered callback!");
//		CallBackInterface cbSingle1 = new CallBack(2356);
//		tcstub.registerCallback(cbSingle1);
//		System.out.println("Registered callback!");
//		CallBackInterface cbSingle2 = new CallBack(2360);
//		tcstub.registerCallback(cbSingle2);
//		System.out.println("Registered callback!");
		/*CallBackInterface cb1 = new CallBack(23);
		CallBackInterface cb2 = new CallBack(33);
		CallBackInterface cb3 = new CallBack(43);
		CallBackInterface[] cb = { cb1, cb2, cb3 };
		tcstub.registerCallbackArray(cb);
		List<CallBackInterface> cblist = new ArrayList<CallBackInterface>();
		CallBackInterface cb4 = new CallBack(53); cblist.add(cb4);
		CallBackInterface cb5 = new CallBack(63); cblist.add(cb5);
		CallBackInterface cb6 = new CallBack(73); cblist.add(cb6);
		tcstub.registerCallbackList(cblist);*/
		/*Enum[] enArr = { Enum.APPLE, Enum.ORANGE, Enum.APPLE, Enum.GRAPE };
		List<Enum> enArr2 = new ArrayList(Arrays.asList(enArr));
		List<Enum> resArr2 = tcstub.handleEnumList(enArr2);
		System.out.println("Enum members: " + resArr2.toString());*/
		Struct str = new Struct();
		str.name = "Rahmadi";
		str.value = 0.123f;
		str.year = 2016;
		Struct str2 = new Struct();
		str2.name = "Trimananda";
		str2.value = 0.124f;
		str2.year = 2017;
		Struct[] arrStr = { str, str2 };
		Struct[] arrRet = tcstub.handleStructArray(arrStr);
		for(Struct st : arrRet) {
			System.out.println("Name: " + st.name);
			System.out.println("Value: " + st.value);
			System.out.println("Year: " + st.year);
		}

//		System.out.println("Return value from callback 1: " + tcstub.callBack() + "\n\n");
//		System.out.println("\n\nCalling short one more time value: " + tcstub.getShort((short)4576) + "\n\n");
//		System.out.println("Return value from callback 2: " + tcstub.callBack() + "\n\n");
//		System.out.println("\n\nCalling short one more time value: " + tcstub.getShort((short)1233) + "\n\n");
//		System.out.println("\n\nCalling short one more time value: " + tcstub.getShort((short)1321) + "\n\n");
		while(true) {}
	}
}

File: Sentinel/iotjava/iotrmi/C++/basics/TestClassCallbacks_Stub.cpp
#include <iostream>
#include <string>
#include "TestClassComplete_Stub.hpp"
#include "CallBack.hpp"

#include "CallBackInterface_Skeleton.cpp"

using namespace std;


TestClassComplete_Stub::TestClassComplete_Stub(int _port, const char* _skeletonAddress, string _callbackAddress, int _rev, bool* _bResult, vector<int> _ports) {
	callbackAddress = _callbackAddress;
	ports = _ports;
	rmiCall = new IoTRMICall(_port, _skeletonAddress, _rev, _bResult);
	set0Allowed.insert(-9998);
	//thread th1 (&TestClassComplete_Stub::___initCallBack, this);
	//th1.detach();
	___regCB();
}

TestClassComplete_Stub::TestClassComplete_Stub(IoTRMICall* _rmiCall, string _callbackAddress, int _objIdCnt, vector<int> _ports) {
	callbackAddress = _callbackAddress;
	rmiCall = _rmiCall;
	objIdCnt = _objIdCnt;
	set0Allowed.insert(-9998);
	//thread th1 (&TestClassComplete_Stub::___initCallBack, this);
	//th1.detach();
	___regCB();
}

TestClassComplete_Stub::~TestClassComplete_Stub() {
	if (rmiCall != NULL) {
		delete rmiCall;
		rmiCall = NULL;
	}
	if (rmiObj != NULL) {
		delete rmiObj;
		rmiObj = NULL;
	}
	for(CallBackInterface* cb : vecCallbackObj) {
		delete cb;
		cb = NULL;
	}
}

void TestClassComplete_Stub::registerCallback(CallBackInterface* _cb) { 
	//CallBackInterface_CallbackSkeleton* skel0 = new CallBackInterface_CallbackSkeleton(_cb, callbackAddress, objIdCnt++);
	CallBackInterface_Skeleton* skel0 = new CallBackInterface_Skeleton(_cb, callbackAddress, objIdCnt++);
	vecCallbackObj.push_back(skel0);
	int ___paramCB0 = 1;
	int methodId = 1;
	string retType = "void";
	int numParam = 1;
	string paramCls[] = { "int" };
	void* paramObj[] = { &___paramCB0 };
	void* retObj = NULL;
	rmiCall->remoteCall(objectId, methodId, retType, paramCls, paramObj, numParam, retObj);
}

void TestClassComplete_Stub::___regCB() {
	int numParam = 3;
	int methodId = -9999;
	string retType = "void";
	string paramCls[] = { "int*", "String", "int" };
	int rev = 0;
	void* paramObj[] = { &ports, &callbackAddress, &rev };
	void* retObj = NULL;
	rmiCall->remoteCall(objectId, methodId, retType, paramCls, paramObj, numParam, retObj);
}

short TestClassComplete_Stub::getShort(short in) { 
	int methodId = 0;
	string retType = "short";
	int numParam = 1;
	string paramCls[] = { "short" };
	void* paramObj[] = { &in };
	short retVal = 0;
	void* retObj = &retVal;
	rmiCall->remoteCall(objectId, methodId, retType, paramCls, paramObj, numParam, retObj);
	return retVal;
}

int TestClassComplete_Stub::callBack() { 
	int methodId = 2;
	string retType = "int";
	int numParam = 0;
	string paramCls[] = {  };
	void* paramObj[] = {  };
	int retVal = 0;
	void* retObj = &retVal;
	rmiCall->remoteCall(objectId, methodId, retType, paramCls, paramObj, numParam, retObj);
	return retVal;
}


int main(int argc, char *argv[])
{

	int port = 5010;
	const char* address = "localhost";
	//const char* address = "192.168.2.191";	// RPi2
	//const char* skeletonAddress = "128.195.136.170";	// dc-9.calit2.uci.edu
	const char* skeletonAddress = "128.195.204.132";
	const char* callbackAddress = "128.195.204.132";	// dw-2.eecs.uci.edu (this machine)
	//const char* skeletonAddress = "192.168.2.108";	// RPi1
	//const char* callbackAddress = "192.168.2.191";	// RPi2
	int rev = 0;
	bool bResult = false;
	vector<int> ports;
	ports.push_back(12345);
	ports.push_back(22346);
	//ports.push_back(32344);
	//ports.push_back(43212);

	TestClassComplete *tcStub = new TestClassComplete_Stub(port, skeletonAddress, callbackAddress, rev, &bResult, ports);
	cout << "==== CALLBACK ====" << endl;
	CallBackInterface *cbSingle = new CallBack(2354);
	tcStub->registerCallback(cbSingle);
	cout << "Return value from callback: " << tcStub->callBack() << endl;

	return 0;
}
